src/main/java/com/couchbase/lite/store/ForestDBStore.java

```diff
@@ -644,105 +648,147 @@ public int compare(String id1, String id2) {
-                }
-                itr = forest.iterator(startKey, endKey, skip, iteratorFlags);
-            }
-            Document doc;
-            while ((doc = itr.nextDocument()) != null) {
+        if (options.getKeys() != null) {
+            rows = new ArrayList<QueryRow>();
+            iteratorFlags |= IteratorFlags.kIncludeDeleted;
+            int total = options.getKeys().size();
+            int read = 0;
+            while (total > 0) {
+                int plan = Math.min(total, MAX_RECORDS_TO_READ_FROM_FORESTDB_AT_ONCE);
+                String[] docIDs = options.getKeys().subList(read, read + plan).toArray(new String[plan]);
                 try {
-                    String docID = doc.getDocID();
-                    if (!doc.exists()) {
-                        Log.v(TAG, "AllDocs: No such row with key=\"%s\"", docID);
-                        QueryRow row = new QueryRow(null, 0, docID, null, null);
-                        rows.add(row);
-                        continue;
-                    }
-
-                    boolean deleted = doc.deleted();
-                    if (deleted &&
-                            options.getAllDocsMode() != Query.AllDocsMode.INCLUDE_DELETED &&
-                            options.getKeys() == null)
-                        continue; // skip deleted doc
-                    if (!doc.conflicted() &&
-                            options.getAllDocsMode() == Query.AllDocsMode.ONLY_CONFLICTS)
-                        continue; // skip non-conflicted doc
-
-                    String revID = doc.getSelectedRevID();
-                    long sequence = doc.getSelectedSequence();
-
-                    RevisionInternal docRevision = null;
-                    if (includeDocs) {
-                        // Fill in the document contents:
-                        docRevision = ForestBridge.revisionObject(doc, docID, revID, true);
-                        if (docRevision == null)
-                            Log.w(TAG, "AllDocs: Unable to read body of doc %s", docID);
-                    }
-
-                    List<String> conflicts = new ArrayList<String>();
-                    if ((options.getAllDocsMode() == Query.AllDocsMode.SHOW_CONFLICTS
-                            || options.getAllDocsMode() == Query.AllDocsMode.ONLY_CONFLICTS)
-                            && doc.conflicted()) {
-                        conflicts = ForestBridge.getCurrentRevisionIDs(doc);
-                        if (conflicts != null && conflicts.size() == 1)
-                            conflicts = null;
+                    DocumentIterator itr = forest.iterator(docIDs, iteratorFlags);
+                    try {
+                        List<QueryRow> retRows = readFromIterator(itr, options, includeDocs, filter, limit);
+                        rows.addAll(retRows);
+                        limit -= retRows.size();
+                        total -= retRows.size();
```

Comment:
> In this case, the row should have missing doc.

What is about the case that the document was deleted? Will it is listed as missing or deleted row?